### PR TITLE
Take into account the excludePrefix in the SDK

### PR DESCRIPTION
**Pull request by rdimitrov**
## Motivation and Context

Issue #841

## How Has This Been Tested?

- [ ] Covered by existing integration testing
- [ ] Added integration testing to cover
- [ ] Tested locally
- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by rdimitrov** [comment]
@lobkovilya I was wondering if you can review my changes as based on the integration tests, I'm not sure if it's better to return directly or raise an error and handle it differently when a interesect fault occurs?

**Comment by lobkovilya** [comment]
@rdimitrov I suppose we can pass `excludedPrefixes` even deeper, right to the `ExtractPrefixes()` function. This approach let us affect on algorithm of extracting prefix and not "give up" so fast and return error :) What do you think about it?

File: controlplane/pkg/prefix_pool/prefixpool.go [hunk 1]
```diff
@@ -47,14 +49,15 @@ func NewPrefixPool(prefixes ...string) (PrefixPool, error) {
 	}, nil
 }
 
-func (impl *prefixPool) Extract(connectionId string, family connectioncontext.IpFamily_Family, requests ...*connectioncontext.ExtraPrefixRequest) (srcIP *net.IPNet, dstIP *net.IPNet, requested []string, err error) {
+func (impl *prefixPool) Extract(connectionId string, family connectioncontext.IpFamily_Family, excludedPrefixes []string, requests ...*connectioncontext.ExtraPrefixRequest) (srcIP *net.IPNet, dstIP *net.IPNet, requested []string, err error) {
```
Comment (lobkovilya) [inline controlplane/pkg/prefix_pool/prefixpool.go]: Probably it's better to create new method
```
ExtractWithExcludedPrefix(connectionId string, family connectioncontext.IpFamily_Family, excludedPrefixes []string, requests ...*connectioncontext.ExtraPrefixRequest) (srcIP *net.IPNet, dstIP *net.IPNet, requested []string, err error)
```
And leave `Extract` as it is:
```
Extract(connectionId string, family connectioncontext.IpFamily_Family, requests ...*connectioncontext.ExtraPrefixRequest) (srcIP *net.IPNet, dstIP *net.IPNet, requested []string, err error) {
    return ExtractWithExcludedPrefix(connectionId, family, []string{}, requests...)
}
```

**Comment by lobkovilya** [comment]
And yeah, test TestExcludePrefixCheck should be changed. For example we could grep logs from icmp-responder and wait for error in logs "IPAM: Allocated IP intersects with excludedPrefix" instead of waiting logs in nsc. Or for that test we can deploy "dump" icmp-responder without excluded prefix check and create one more test to check logic that you have provided to sdk

File: controlplane/pkg/prefix_pool/prefixpool.go [hunk 2]
```diff
@@ -201,13 +205,44 @@ func ExtractPrefixes(prefixes []string, requests ...*connectioncontext.ExtraPref
...
+	}
+
+	if intersect {
+		return fmt.Errorf("IPAM: Allocated IP intersects with excludedPrefix")
+	}
+
+	return nil
+}
+
 func ExtractPrefix(prefixes []string, prefixLen uint32) (string, []string, error) {
```
Comment (lobkovilya) [inline controlplane/pkg/prefix_pool/prefixpool.go]: This is function where choosing of appropriate prefix happen. So we need to pass excludedPrefixes here, to affect on this algorithm.

There is a cycle over all available prefixes. Firstly it tries to find prefix with length equal to `prefixLen`. In case of absent of such prefix we find a candidate to split in 2 subnet. So please provide additional logic of choosing that candidate - it mustn't be equal to one of excludedPrefix or be included in one of excludedPrefix

**Comment by lobkovilya** [comment]
It looks good, but I assume that we still have situations when we have available prefix, but algorithm returns error. 

Imagine that we call:
```
ExtractPrefixes(string[]{"10.20.0.0/16"}, 30, string[]{"10.20.0.0/24"})
```
Cycle https://github.com/networkservicemesh/networkservicemesh/pull/888/files#diff-32a094c1e510859d935b4e71d8de9103R278 goes down the tree of prefixes and always choose left branch (subnet with last prefix bit equals to 0). Cycle stops when size of `rootCIDRNet` will equal to `prefixLen`. So we easily achieve `10.20.0.0/30` but final check at https://github.com/networkservicemesh/networkservicemesh/pull/888/files#diff-32a094c1e510859d935b4e71d8de9103R301 possibly says that `10.20.0.0/30` intersects with `10.20.0.0/24` and it leads to error of `ExtractPrefixes` function. But obviously we could find subnet of `10.20.0.0/16` with size 30 that is not intersected with `excludedPrefixes`.

I haven't had a chance to check it in running program, but I suspect that provided example is valid. Could you please check it?

**Comment by rdimitrov** [comment]
Appreciate your feedback, @lobkovilya 👍 I created a new test for that case and it does fail, so I'll update it 👍

File: controlplane/pkg/prefix_pool/prefixpool.go [hunk 3]
```diff
@@ -201,13 +205,38 @@ func ExtractPrefixes(prefixes []string, requests ...*connectioncontext.ExtraPref
...
+	}
+
+	if intersect {
+		return fmt.Errorf("IPAM: Allocated IP intersects with excludedPrefix")
+	}
+
+	return nil
+}
+
+func ExtractPrefix(prefixes []string, prefixLen uint32, excludedPrefixes []string) (string, []string, error) {
```
Comment (rdimitrov) [inline controlplane/pkg/prefix_pool/prefixpool.go]: @lobkovilya Sorry for taking a bit longer to reply. I did try the test you suggested and although currently it is handled as a false negative (stating that it cannot allocate the desired network when in reality it is possible), digging deeper seems that the address allocation algorithm (the one with the subnet splitting) doesn't proceed to check the range from 10.20.1.0 to 10.20.255.255 as it should, but it ends at the 10.20.0.0 one.

File: controlplane/pkg/prefix_pool/prefixpool.go [hunk 4]
```diff
@@ -221,24 +250,29 @@ func ExtractPrefix(prefixes []string, prefixLen uint32) (string, []string, error
...
 			resultPrefixes := append(prefixes[:idx], prefixes[idx+1:]...)
+			/* Verify the allocated addresses does not intersect with one from the excludedPrefix list */
+			err := verifyExcluding(resultPrefix, excludedPrefixes)
+			if err != nil {
+				return "", prefixes, err
+			}
 			// Lets remove from list and return
 			return resultPrefix, resultPrefixes, nil
 		} else {
 			// Update minimal root for split
```
Comment (haiodo) [inline controlplane/pkg/prefix_pool/prefixpool.go]: Hi, We also need to choose a candidate to split here with verification subnet is fit based on 
verifyExcluding(resultPrefix, excludedPrefixes)

File: controlplane/pkg/prefix_pool/prefixpool.go [hunk 5]
```diff
@@ -221,24 +250,29 @@ func ExtractPrefix(prefixes []string, prefixLen uint32) (string, []string, error
...
 		return "", prefixes, fmt.Errorf("Failed to find room to have prefix len %d at %v", prefixLen, prefixes)
 	}
 
-	resultPrefixRoot := prefixes[max_prefix_idx]
-	right_parts := []string{}
+	resultPrefixRoot := prefixes[maxPrefixIndex]
+	rightParts := []string{}
 
 	_, rootCIDRNet, _ := net.ParseCIDR(resultPrefixRoot)
 	for {
```
Comment (haiodo) [inline controlplane/pkg/prefix_pool/prefixpool.go]: Here we need to check and choose a direction we need to split based on 
verifyExcluding(resultPrefix, excludedPrefixes) to select left or right subtree, so we will not split just to left one.

File: controlplane/pkg/prefix_pool/prefixpool.go [hunk 6]
```diff
@@ -256,15 +290,18 @@ func ExtractPrefix(prefixes []string, prefixLen uint32) (string, []string, error
...
-	resultPrefixes = append(resultPrefixes, reverse(right_parts)...)
-	resultPrefixes = append(resultPrefixes, prefixes[max_prefix_idx+1:]...)
-
-	// return result
+	var resultPrefixes []string
+	resultPrefixes = append(resultPrefixes, prefixes[:maxPrefixIndex]...)
+	resultPrefixes = append(resultPrefixes, reverse(rightParts)...)
+	resultPrefixes = append(resultPrefixes, prefixes[maxPrefixIndex+1:]...)
+	/* Verify the allocated addresses does not intersect with one from the excludedPrefix list */
+	err := verifyExcluding(rootCIDRNet.String(), excludedPrefixes)
```
Comment (haiodo) [inline controlplane/pkg/prefix_pool/prefixpool.go]: This check is not required since we will already choose a matched subtree.

**Comment by nickolaev** [comment]
@haiodo can you please confirm the current state if the PR is in line with your vision of it?

File: .circleci/config.yml [hunk 7]
```diff
@@ -307,23 +307,23 @@ workflows:
           requires:
             - packet-deploy
       - integration-tests:
-          name: "packet test basic usecase"
+          name: "packet-test-basic-usecase"
```
Comment (nickolaev) [inline .circleci/config.yml]: Makes sense but please do it in a separate PR.

File: controlplane/pkg/nsm/nsm.go [hunk 8]
```diff
@@ -389,8 +390,8 @@ func (srv *networkServiceManager) isLocalEndpoint(endpoint *registry.NSERegistra
 }
 
 /**
-	ctx - we assume it is big enought to perform connection.
- */
+ctx - we assume it is big enought to perform connection.
```
Comment (nickolaev) [inline controlplane/pkg/nsm/nsm.go]: Please put these in a separate PR.

File: controlplane/pkg/prefix_pool/prefixpool.go (hunk 1 above)
Comment (nickolaev) [reply controlplane/pkg/prefix_pool/prefixpool.go]: @rdimitrov, please share your approach here. Why is it better?

File: controlplane/pkg/prefix_pool/prefixpool_test.go [hunk 9]
```diff
@@ -127,12 +140,29 @@ func TestExtract2_ipv6(t *testing.T) {
...
 
+func TestExtractExclude(t *testing.T) {
+	RegisterTestingT(t)
+
+	prefix, _, err := ExtractPrefix([]string{"10.20.0.0/16"}, 30, []string{"10.20.0.0/24"})
+	Expect(err).To(BeNil())
+	Expect(prefix).To(Equal("10.20.128.0/30"))
+}
+
+func TestExtractExcludeIntersect(t *testing.T) {
```
Comment (nickolaev) [inline controlplane/pkg/prefix_pool/prefixpool_test.go]: Are we sure we are covering all the needed cases? Maybe we need more tests?

File: controlplane/pkg/tests/nsmd_remote_test.go [hunk 10]
```diff
@@ -215,27 +215,27 @@ func TestNSMDDelayRemoteMechanisms(t *testing.T) {
 
 	type Response struct {
 		nsmResponse *connection.Connection
-		err error
+		err         error
```
Comment (nickolaev) [inline controlplane/pkg/tests/nsmd_remote_test.go]: The changes in this file are not relevant to the patch.Please consider a separate PR.

File: dataplane/pkg/common/egress_interface.go [hunk 11]
```diff
@@ -36,9 +36,9 @@ type EgressInterface interface {
```
Comment (nickolaev) [inline dataplane/pkg/common/egress_interface.go]: Not relevant. Conseder a seaprate PR.

File: dataplane/pkg/common/egress_interface_test.go [hunk 12]
```diff
@@ -1,4 +1,4 @@
-package  common
```
Comment (nickolaev) [inline dataplane/pkg/common/egress_interface_test.go]: Not relevant. Please consider a separate PR.

File: dataplane/vppagent/pkg/converter/acl_converter.go [hunk 13]
```diff
@@ -105,7 +105,7 @@ func getIcmp(parsed map[string]string) (*vpp_acl.ACL_Rule_IpRule_Icmp, error) {
 	if err != nil {
```
Comment (nickolaev) [inline dataplane/vppagent/pkg/converter/acl_converter.go]: Separate PR?

File: dataplane/vppagent/pkg/converter/kernel_connection_converter.go [hunk 14]
```diff
@@ -91,11 +91,11 @@ func (c *KernelConnectionConverter) ToDataRequest(rv *configurator.Config, conne
 			Enabled:     true,
```
Comment (nickolaev) [inline dataplane/vppagent/pkg/converter/kernel_connection_converter.go]: Separate PR?

File: dataplane/vppagent/pkg/converter/memif_interface_converter.go [hunk 15]
```diff
@@ -20,9 +20,9 @@ import (
 	"github.com/ligato/vpp-agent/api/models/vpp"
```
Comment (nickolaev) [inline dataplane/vppagent/pkg/converter/memif_interface_converter.go]: Separate PR?

File: dataplane/vppagent/pkg/converter/remote_connection_converter.go [hunk 16]
```diff
@@ -82,7 +82,7 @@ func (c *RemoteConnectionConverter) ToDataRequest(rv *configurator.Config, conne
 			Vxlan: &vpp_interfaces.VxlanLink{
```
Comment (nickolaev) [inline dataplane/vppagent/pkg/converter/remote_connection_converter.go]: Separate PR?

File: dataplane/vppagent/pkg/vppagent/vppagent.go [hunk 17]
```diff
@@ -16,15 +16,16 @@ package vppagent
```
Comment (nickolaev) [inline dataplane/vppagent/pkg/vppagent/vppagent.go]: Separate PR?

File: examples/cmd/vppagent-firewall-nse/vppagent.go [hunk 18]
```diff
@@ -58,7 +58,7 @@ func (vxc *vppAgentXConnComposite) crossConnecVppInterfaces(ctx context.Context,
 	}
```
Comment (nickolaev) [inline examples/cmd/vppagent-firewall-nse/vppagent.go]: Separate PR?

File: examples/cmd/vppagent-icmp-responder-nse/vppagent.go [hunk 19]
```diff
@@ -45,9 +45,9 @@ func (ns *vppagentComposite) CreateVppInterface(ctx context.Context, nseConnecti
 		return err
```
Comment (nickolaev) [inline examples/cmd/vppagent-icmp-responder-nse/vppagent.go]: Separate PR?

File: examples/cmd/vppagent-nsc/vppagent.go [hunk 20]
```diff
@@ -70,8 +70,7 @@ func Reset(vppAgentEndpoint string) error {
 	client := configurator.NewConfiguratorClient(conn)
```
Comment (nickolaev) [inline examples/cmd/vppagent-nsc/vppagent.go]: Separate PR?

File: pkg/tools/tools.go [hunk 21]
```diff
@@ -126,7 +126,7 @@ func WaitForPortAvailable(ctx context.Context, protoType string, registryAddress
 		default:
```
Comment (nickolaev) [inline pkg/tools/tools.go]: Separate PR?

File: scripts/aws/create-kubernetes-cluster.go [hunk 22]
```diff
@@ -22,15 +22,15 @@ import (
 )
```
Comment (nickolaev) [inline scripts/aws/create-kubernetes-cluster.go]: Separate PR?

File: scripts/aws/cubernetes-cluster.go [hunk 23]
```diff
@@ -5,26 +5,26 @@ import (
 	"os"
```
Comment (nickolaev) [inline scripts/aws/cubernetes-cluster.go]: Separate PR?

File: scripts/aws/destroy-kubernetes-cluster.go [hunk 24]
```diff
@@ -45,24 +45,26 @@ func DeleteEksRole(iamClient *iam.IAM, eksRoleName *string) {
```
Comment (nickolaev) [inline scripts/aws/destroy-kubernetes-cluster.go]: Separate PR?

File: scripts/prepare-circle-integration-tests.sh [hunk 25]
```diff
@@ -17,7 +17,7 @@
 
 set -xe
 
-KUBECTL_VERSION=v1.13.3
+KUBECTL_VERSION=v1.13.4
```
Comment (nickolaev) [inline scripts/prepare-circle-integration-tests.sh]: That does not belong here.

File: test/kube_testing/pods/nsmd.go [hunk 26]
```diff
@@ -118,7 +118,7 @@ func NSMgrPodWithConfig(name string, node *v1.Node, config *NSMgrPodConfig) *v1.
 					Image:           "networkservicemesh/nsmdp",
```
Comment (nickolaev) [inline test/kube_testing/pods/nsmd.go]: Separate PR.

File: test/kube_testing/pods/vppagent_dataplane.go [hunk 27]
```diff
@@ -89,7 +89,7 @@ func createVPPDataplanePod(name string, node *v1.Node, liveness, readiness *v1.P
 					},
```
Comment (nickolaev) [inline test/kube_testing/pods/vppagent_dataplane.go]: Separate PR.

**Comment by lobkovilya** [comment]
Seems like current implementation still has "false negative" behavior. 

Algorithm has 2 phases: election of appropriate candidate for splitting and splitting itself. Now election phase uses so called "best-fit" strategy:
```go
for idx, prefix := range prefixes {
    ...
    if uint32(parentLen) < prefixLen && (parentLen > maxPrefix || maxPrefix == 0) {
	maxPrefix = parentLen
	maxPrefixIndex = idx
    }
}
```
Since there is no check whether candidate intersects with `excludePrefixes` or not, we can't unambiguously say that it is possible to extract prefixes from it. In other words another candidate could be "wider" (has shorter prefix) but without intersections with `excludePrefixes`. 

But simple reelection of candidate after Phase 2 ("splitting") failed will not help. Turns out splitting is harder than we thought before. Checking which subtree intersects with `excludePrefixes` and going down to opposite subtree will not work in case of `excludePrefixes` consist of several non-overlapping prefixes with different sizes. Imagine that candidate has prefix `/16`, `excludePrefixes` contains couple tiny `/30` subnets. One of subnet included to left `/17` subtree and other to right. Algorithm will go down to the left subtree and may fail to find appropriate prefix while right subtree could contain required prefix.

I leave some draft description of possible algorithm in related issue, hope it would be helpful.

File: controlplane/pkg/prefix_pool/prefixpool.go [hunk 28]
```diff
@@ -47,6 +51,121 @@ func NewPrefixPool(prefixes ...string) (PrefixPool, error) {
...
+
+		/* 1. Check if each excluded entry overlaps with the available prefix */
+		for _, prefix := range copyPrefixes {
+			intersecting := false
+			excludedIsBigger := false
+			_, subnetPrefix, _ := net.ParseCIDR(prefix)
+			intersecting, excludedIsBigger = intersect(subnetExclude, subnetPrefix)
+			/* 1.1. If intersecting, check which one is bigger */
+			if intersecting {
+				/* 1.1.1. If excluded is bigger, we remove the original entry */
```
Comment (lobkovilya) [inline controlplane/pkg/prefix_pool/prefixpool.go]: @rdimitrov Hm, probably I am mistaken, but if we simply remove original entry in that case, then during releasing we increase the original size of prefix pool. I mean the following test most likely will fail:
```go
pool, err := NewPrefixPool("10.20.0.0/16", "2.20.0.0/16")
Expect(err).To(BeNil())
excludedPrefix := []string{"2.20.0.0/8"}

excluded, err := pool.ExcludePrefixes(excludedPrefix)

Expect(err).To(BeNil())
Expect(pool.GetPrefixes()).To(Equal([]string{"10.20.0.0/16"}))

err = pool.ReleaseExcludedPrefixes(excluded)
Expect(err).To(BeNil())
Expect(pool.GetPrefixes()).To(Equal([]string{"10.20.0.0/16", "2.20.0.0/16"}))
```

File: sdk/endpoint/ipam.go [hunk 29]
```diff
@@ -51,12 +51,24 @@ func (ice *IpamEndpoint) Request(ctx context.Context, request *networkservice.Ne
 		return nil, err
 	}
 
+	/* Exclude the prefixes from the pool of available prefixes */
+	excludedPrefixes, err := ice.prefixPool.ExcludePrefixes(request.Connection.Context.ExcludedPrefixes)
```
Comment (lobkovilya) [inline sdk/endpoint/ipam.go]: May be it is better to save copy of prefixPool before excluding prefixes

File: sdk/endpoint/ipam.go [hunk 30]
```diff
@@ -51,12 +51,24 @@ func (ice *IpamEndpoint) Request(ctx context.Context, request *networkservice.Ne
...
+	}
+
 	//TODO: We need to somehow support IPv6.
 	srcIP, dstIP, prefixes, err := ice.prefixPool.Extract(request.Connection.Id, connectioncontext.IpFamily_IPV4, request.Connection.Context.ExtraPrefixRequest...)
 	if err != nil {
 		return nil, err
 	}
 
+	/* Release the actual prefixes that were excluded during IPAM */
+	err = ice.prefixPool.ReleaseExcludedPrefixes(excludedPrefixes)
```
Comment (lobkovilya) [inline sdk/endpoint/ipam.go]: And restore that copy right here? 
Then we can avoid implementing `ReleaseExcludePrefixes` method. What do you think about it?

Comment (rdimitrov) [reply sdk/endpoint/ipam.go]: Make sense, but if we restore the original state of `prefixes` at this point - this happens right after we do the IPAM using the altered copy with the excluded prefixes. If successful, this results in extracting prefixes from it, so if we do that, in the end, we should have the troubles to take into account that change and translate it back to the original, unaltered copy.

File: controlplane/pkg/prefix_pool/prefixpool.go (hunk 28 above)
Comment (rdimitrov) [reply controlplane/pkg/prefix_pool/prefixpool.go]: Thanks, @lobkovilya! 👍 I've updated the tests with that use case and added a fix for it.

Comment (rdimitrov) [reply controlplane/pkg/prefix_pool/prefixpool.go]: By the way, I found this one to be really useful for verifying what to expect - https://www.techzoom.net/Tools/IPAddressCalculator 👍

**Comment by rdimitrov** [comment]
@nickolaev I assume the "requested changes" part was already covered

File: sdk/endpoint/ipam.go (hunk 30 above)
Comment (lobkovilya) [reply sdk/endpoint/ipam.go]: Yep, you're absolutely right!
